compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1106,7 +1113,129 @@ impl<'a> Parser<'a> {
                 e
             })?;
 
-        let enum_definition = EnumDef { variants: variants.into_iter().flatten().collect() };
+        let enum_definition =
+            EnumDef { variants: variants.into_iter().filter_map(|v| v).collect() };
+
+        // Implicitly add a #[derive(::core::enums::AsRepr)] attribute to all fieldless enums with an
+        // explicit repr.
```

Comment:
Instead of implicitly adding this derive in the parser, I think a better approach would be to either have an ast pass that directly writes the expanded form or do this during hir lowering. It should also use lang items instead of hard coding paths. Lang items are not forge-able using stable rust, `::core` is using `--extern core=...`, so using a hard coded path would make it possible to depend on this implementation detail, which must be impossible.

Reply:
Agree with this comment, but if for some reason we find we can't do that (which I don't expect), we should at least move this to its own method.